hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/HoodieIndexUtils.java

```diff
@@ -310,6 +312,56 @@ public static <R> HoodieData<HoodieRecord<R>> mergeForPartitionUpdates(
             return Arrays.asList(deleteRecord, getTaggedRecord(merged, Option.empty())).iterator();
           }
         });
-    return taggedUpdatingRecords.union(newRecords);
+    return taggedUpdatingRecords.union(taggedNewRecords);
+  }
+
+  public static <R> HoodieData<HoodieRecord<R>> tagGlobalLocationBackToRecords(
+      HoodieData<HoodieRecord<R>> incomingRecords,
+      HoodiePairData<String, HoodieRecordGlobalLocation> keyAndExistingLocations,
+      boolean mayContainDuplicateLookup,
+      boolean shouldUpdatePartitionPath,
+      HoodieWriteConfig config,
+      HoodieTable table) {
+    final HoodieRecordMerger merger = config.getRecordMerger();
+
+    HoodiePairData<String, HoodieRecord<R>> keyAndIncomingRecords =
+        incomingRecords.mapToPair(record -> Pair.of(record.getRecordKey(), record));
+
+    // Pair of incoming record and the global location if meant for merged lookup in later stage
+    HoodieData<Pair<HoodieRecord<R>, Option<HoodieRecordGlobalLocation>>> incomingRecordsAndLocations
+        = keyAndIncomingRecords.leftOuterJoin(keyAndExistingLocations).values()
+        .map(v -> {
+          final HoodieRecord<R> incomingRecord = v.getLeft();
+          Option<HoodieRecordGlobalLocation> currentLocOpt = Option.ofNullable(v.getRight().orElse(null));
+          if (currentLocOpt.isPresent()) {
+            HoodieRecordGlobalLocation currentLoc = currentLocOpt.get();
+            boolean shouldPerformMergedLookUp = mayContainDuplicateLookup
+                || !Objects.equals(incomingRecord.getPartitionPath(), currentLoc.getPartitionPath());
+            if (shouldUpdatePartitionPath && shouldPerformMergedLookUp) {
+              return Pair.of(incomingRecord, currentLocOpt);
+            } else {
+              // - When update partition path is set to false,
+              //   the incoming record will be tagged to the existing record's partition regardless of being equal or not.
+              // - When update partition path is set to true,
+              //   the incoming record will be tagged to the existing record's partition
+              //   when partition is not updated and the look-up won't have duplicates (e.g. COW, or using RLI).
+              return Pair.of((HoodieRecord<R>) getTaggedRecord(
```

Comment:
should this be 
```
createNewHoodieRecord(incomingRecord, currentLoc, merger), Option.empty())
```

i.e. last arg as Option.empty. so that within mergeForPartitionUpdatesIfNeeded, we don't need to load those partition keys.

Reply:
new record creation needs optimization; i have not finished it yet.

Reply:
got it

Reply:
refactored